File: 1889-minimum-space-wasted-from-packaging/1889-minimum-space-wasted-from-packaging.cpp
class Solution {
public:
    int minWastedSpace(vector<int>& packages, vector<vector<int>>& boxes) {
        sort(packages.begin(), packages.end());
        long res = LONG_MAX, m = 1e9 + 7, sum = 0;
        
        for(int p : packages) sum += p;
        
        for(auto& b : boxes) {
            sort(b.begin(), b.end());
            if(b.back() < packages.back()) continue;
            // int j = 0, i = 0; long cur = 0;
            long cur = 0, i = 0, j;
            for(int b1 : b) {
                j = upper_bound(packages.begin(), packages.end(), b1) - packages.begin();
                cur += (j - i) * b1;
                i = j;
            }
            res = min(res, cur);
        }
        
        return res < LONG_MAX ? (res - sum) % m : -1;
    }
};

File: 1889-minimum-space-wasted-from-packaging/1889-minimum-space-wasted-from-packaging.py
class Solution:
    def minWastedSpace(self, packages: List[int], boxes: List[List[int]]) -> int:
        packages.sort()
        res = float('inf')
        
        for b in boxes:
            b.sort()
            if b[-1] < packages[-1]: continue
            cur = i = 0
            for b1 in b:
                j = bisect.bisect(packages, b1, i)
                cur += (j - i) * b1
                i = j
            res = min(res, cur)
        
        return (res - sum(packages)) % (10 ** 9 + 7) if res < float('inf') else -1